src/DGtal/shapes/SurfaceMesh.ih

```diff
@@ -840,6 +852,223 @@ computeNonManifoldEdges() const
+  const Vertices& rvtx = incidentVertices( rf );
+  if ( rvtx.size() != 3 ) return false;   //< right face is not a triangle
+  const Vertices& lvtx = incidentVertices( lf );
+  if ( lvtx.size() != 3 ) return false;   //< left  face is not a triangle
+
+  // (3) the two other vertices of the quad are not already neighbors.
+  Vertex i, j;
+  std::tie( i, j ) = edgeVertices( e );
+  const auto    ir = ( rvtx[ 0 ] == i ) ? 0 : ( ( rvtx[ 1 ] == i ) ? 1 : 2 );
+  const auto    il = ( lvtx[ 0 ] == i ) ? 0 : ( ( lvtx[ 1 ] == i ) ? 1 : 2 );
+  const Vertex   k = rvtx[ (ir + 1) % 3 ]; // right triangle is (j,i,k)
+  const Vertex   l = lvtx[ (il + 2) % 3 ]; // left  triangle is (i,j,l). 
+  if ( ! ( k != i && k != j && l != i && l != j && k != l ) )
+    {
+      /// May happen if we have identical triangles (i,j,k) and (j,i,k)
+      if ( k == l ) return false;
+      /// Otherwise there is a problem.
+      trace.error() << "[SurfaceMesh::isFlippable] Invalid neighbors to edge: "
+		    << "(i,j,k,l) == (" << i << "," << j << "," << k << "," << l << ")"
+		    << std::endl
+		    << "right=(" << rvtx[ 0 ] << "," << rvtx[ 1 ] << "," << rvtx[ 2 ] << ")" << std::endl
+		    << "left =(" << lvtx[ 0 ] << "," << lvtx[ 1 ] << "," << lvtx[ 2 ] << ")" << std::endl;
+      return false;
+    }
+  const Vertices& Nk = neighborVertices( k );
+  const auto     itl = std::find( Nk.cbegin(), Nk.cend(), l );
+  return itl == Nk.cend();
+}
+
+//-----------------------------------------------------------------------------
+template <typename TRealPoint, typename TRealVector>
+typename DGtal::SurfaceMesh<TRealPoint, TRealVector>::VertexPair
+DGtal::SurfaceMesh<TRealPoint, TRealVector>::
+otherDiagonal( const Edge e ) const
+{
+  // only valid if `isFlippable( e )` is true.
+  const Faces&  rfaces = edgeRightFaces( e );
+  const Faces&  lfaces = edgeLeftFaces ( e );
+  const Face      rf   = rfaces.front();  //< some `(..., j, i, ... )` since faces are ccw.
+  const Face      lf   = lfaces.front();  //< some `(..., i, j, ... )` since faces are ccw.
+  const Vertices& rvtx = incidentVertices( rf );
+  const Vertices& lvtx = incidentVertices( lf );
+  Vertex i, j;
+  std::tie( i, j ) = edgeVertices( e );
+  const auto    ir = ( rvtx[ 0 ] == i ) ? 0 : ( ( rvtx[ 1 ] == i ) ? 1 : 2 );
+  const auto    il = ( lvtx[ 0 ] == i ) ? 0 : ( ( lvtx[ 1 ] == i ) ? 1 : 2 );
+  const Vertex   k = rvtx[ (ir + 1) % 3 ]; // right triangle is (j,i,k)
+  const Vertex   l = lvtx[ (il + 2) % 3 ]; // left  triangle is (i,j,l).
+  return ( k < l ) ? std::make_pair( k, l ) : std::make_pair( l, k );
+}
+
+
+//-----------------------------------------------------------------------------
+template <typename TRealPoint, typename TRealVector>
+void
+DGtal::SurfaceMesh<TRealPoint, TRealVector>::
+flip( const Edge e, bool recompute_face_normals )
+{
+  /*
+           l                   l
+          / \                 /|\
```

Comment:
Maybe this kind of illustration would be nice in the doc. It explains the flip better than any text, imho

Reply:
Ok I try to do that.